mlos_core/mlos_core/optimizers/bayesian_optimizers/smac_optimizer.py

```diff
@@ -0,0 +1,253 @@
+        from smac import HyperparameterOptimizationFacade as Optimizer_Smac
+        from smac import Scenario
+        from smac.intensifier.abstract_intensifier import AbstractIntensifier
+        from smac.initial_design import LatinHypercubeInitialDesign
+        from smac.main.config_selector import ConfigSelector
+        from smac.random_design.probability_design import ProbabilityRandomDesign
+        from smac.runhistory import TrialInfo
+
+        # Store for TrialInfo instances returned by .ask()
+        self.trial_info_map: Mapping[ConfigSpace.Configuration, TrialInfo] = {}
+
+        # The default when not specified is to use a known seed (0) to keep results reproducible.
+        # However, if a `None` seed is explicitly provided, we let a random seed be produced by SMAC.
+        # https://automl.github.io/SMAC3/main/api/smac.scenario.html#smac.scenario.Scenario
+        seed = -1 if seed is None else seed
+
+        # Create temporary directory for SMAC output (if none provided)
+        self.temp_output_directory: Optional[TemporaryDirectory] = None
+        if output_directory is None:
+            # pylint: disable=consider-using-with
+            try:
+                self.temp_output_directory = TemporaryDirectory(ignore_cleanup_errors=True)  # Argument added in Python 3.10
+            except TypeError:
+                self.temp_output_directory = TemporaryDirectory()
+            output_directory = self.temp_output_directory.name
+
+        # Instantiate Scenario
+        scenario: Scenario = Scenario(
+            self.optimizer_parameter_space,
+            name=run_name,
+            output_directory=Path(output_directory),
+            deterministic=True,
+            n_trials=max_trials,
+            seed=seed or -1,  # if -1, SMAC will generate a random seed internally
+            n_workers=1,  # Use a single thread for evaluating trials
+        )
+        intensifier: AbstractIntensifier = Optimizer_Smac.get_intensifier(scenario, max_config_calls=1)
+        config_selector: ConfigSelector = ConfigSelector(scenario, retrain_after=1)
+
+        # Customize SMAC's randomized behavior
+        initial_design: Optional[LatinHypercubeInitialDesign] = None
+        if n_random_init is not None:
+            initial_design = LatinHypercubeInitialDesign(scenario=scenario, n_configs=n_random_init)
+        random_design: Optional[ProbabilityRandomDesign] = None
+        if n_random_probability is not None:
+            random_design = ProbabilityRandomDesign(probability=n_random_probability)
+
+        # Create SMAC optimizer
+        self.base_optimizer = Optimizer_Smac(
+            scenario,
+            SmacOptimizer._dummy_target_func,
+            initial_design=initial_design,
+            intensifier=intensifier,
+            random_design=random_design,
+            config_selector=config_selector,
+            overwrite=True,
+        )
+
+    def __del__(self) -> None:
```

Comment:
I am not sure it's a good idea to rely on `__del__()` to clean up the resources (Python does not guarantee it will be invoked at all). A much better solution would be to use `__enter__()` / `__exit__()` methods to create a `with` context (e.g., please see our own `mlos_bench.storage.base_storage.Storage.Experiment`) 

I understand that this might require changes to the overall optimzier API, and I am ok if we do this in a separate PR - but for now let's create an explicit `.close()` method and call it and also open an issue on github so we don't forget to implement a `with` context later

Reply:
Hmm I wasn't aware that it was possible for `__del__()` to not be called -- nice catch!

I agree that adding a `.close()` method would be the proper way to do it, since relying on a context manager for the optimizer object can be restrictive for the user. cc'ing @bpkroth

Reply:
Just added a `.cleanup()` method to `Optimizer` class, as well as an implementation for `SmacOptimizer` concrete class.

Reply:
Ideally, we need both `.cleanup()` and `__enter__/__exit__`, but having just `.cleanup()` is a good first step. I was proposing `.close()` as a name just because every other class I know has it (so it's like `Closeable` interface in Java). but I'm fine with `.cleanup()`, too